modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/segstore/SegmentInfo.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.raft.storage.segstore;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Arrays;
+
+/**
+ * Information about a segment file stored in a {@link IndexMemTable}.
+ */
+class SegmentInfo {
+    private static class ArrayWithSize {
+        private static final int INITIAL_CAPACITY = 10;
+
+        private final int[] array;
+
+        private final int size;
+
+        ArrayWithSize() {
+            this(new int[INITIAL_CAPACITY], 0);
+        }
+
+        private ArrayWithSize(int[] array, int size) {
+            this.array = array;
+            this.size = size;
+        }
+
+        ArrayWithSize add(int element) {
+            // The array can be shared between multiple instances, but since it always grows and we read at most "size" elements,
+            // we don't need to copy it every time.
+            int[] array = this.array;
+
+            if (size == array.length) {
+                array = Arrays.copyOf(array, array.length * 2);
+            }
+
+            array[size] = element;
+
+            return new ArrayWithSize(array, size + 1);
+        }
+
+        int get(int index) {
```

Comment:
As a suggestion, this method is not needed, it is an inner class, it is at your discretion.

Reply:
I added it on purpose, it looks cleaner than using fields directly